### ISSUE TITLE
How does this compare to wsproto?

### ISSUE THREAD

**Issue by njsmith** [theelous3/noio_ws#1]
Are you aware of [wsproto](https://github.com/python-hyper/wsproto/)? Any comparisons? Interest in joining forces?

**Comment by theelous3** [comment]
'ello guv'nor.

I am aware of it! I took a look a while ago and was disheartened by the lack of docs. I figured it'd probably take me just as long to write a lib myself and document it, than to dive in and figure out the workings of wsproto without knowing it. (I knew absolutely nothing about websockets before diving in to this. Blind leading the blind, and so forth.)

>Any comparisons?

I haven't seen what, for example, an example client looks like in wsproto. I would love to, if you have one handy. I'd actually sort've forgotten about it as I wrote this lib, and going back just now to have a quick scan it looks lovely.

It seems you take a much more integrated approach to the opening handshake phase, where as I, at least for now, keep it at arms length. I'm definitely paying a price for that design decision in terms of the current api. I'm not yet sure if the compartmentalisation is worth it.

Having more familiarity with wsproto, I think you would be better equipped to take a quick scan of a noio_ws client and make a more reasoned off the cuff comparison. Here's a client that connects to a public echo server, sends gibberish, prints the echo response, and closes (curio based). Much of it should be very familiar :)

https://github.com/theelous3/noio_ws/blob/master/tests/junk/echo_client.py

> Interest in joining forces?

Of course! Though wsproto being so much more mature I'm not sure what I could offer. What stuff are you guys working on at the moment? It's insanely late here, but I'll take a closer look at wsproto tomorrow and see what's up.

PS: lol. I just realised how badly I shadowed `type()` in my example. That'll teach me to write things on no sleep.

**Comment by njsmith** [comment]
It's true, the lack of docs is a major gap. A few of us have been poking at getting the project up to speed, but so far more energy has gone into things like tests and rewriting the frame handling code several times... it's hardly *mature*, there's very few people actually using it for anything yet, but it has had a few people's eyes on it at least, and the test suite is now reasonably thorough.

One thing that jumps out at me is that wsproto delivers partial messages as they're available (sort of analogous to how `h11.Data` represents a piece of the request/response body), to avoid arbitrarily having to buffer arbitrarily large amounts of data; I don't see anything similar in noio_ws.

**Comment by theelous3** [comment]
Nope. At least not yet. I'm just not sure if it's worth it. I'm very much developing this and learning about websockts in a bubble, so if you have insight on the matter I'd love to hear it.

My opinions thus far have been:

___

That's what fragmentation is for.

With such a bit-concerned protocol, sending messages is essentially "free". Any client/server not taking advantage of that is insane. I feel like the protocol is set up in such a way that if the fragmenting of fragments is required, the producer of those fragments is doing something horribly wrong.

As long as there's a way to protect against maliciously large messages / message streams then there's little benefit (sometimes no benefit) to feeding the payload out in smaller chunks. The underlying network behaves the same but your application gets the data faster. If the application needs a whole frame it has to wait anyway and deal with the increased complexity of reassembling the frame. If it doesn't need the whole frame, then it's a win for the extra-fragmentation.

As far as I could tell all of this is only when there are [no extensions present](https://tools.ietf.org/html/rfc6455#section-5.4) anyway. Deflate and the likes require a full frame, as could many subprotocols etc.

For example, slack uses un-sub-fragmentable<sup>1</sup> json based payloads, and from poking around I've gathered this is popular, and that relatively light payloads are the use case for websockets. If someone is using single websocket frames to send 4gb movies or whatever, then they're simply using the wrong protocol and whatever poor service is on the other end should reject them. (Why the protocol allows for multiple exabyte sized frames is beyond me.)

<sup>1</sup>_(Not really un-sub-fragmentable, but pointless to do so.)_
___
Would be very happy to hear dissenting opinion on the above. You say the lack of the feature jumps out at you so I take it you guys see it as quite important. I feel the importance is _whooshing_ me atm.

If it's the correct question to ask; in your opinion is the immediacy of data availability to the application worth any increased complexity? Furthermore, have you come across run-of-the-mill websocket based services that require huge amounts of memory per frame when not sub-fragmented?

**Comment by njsmith** [comment]
Eh, it jumps out at me because it's a bit subtle and tricky and I was just reviewing some changes to wsproto's handling of it :-). But I do think it's the right thing to do: if the lower layer sends back fragments then the upper layer can define whatever coalescing policy it pleases; but if the lower layer does coalescing, the upper layer is stuck with it. Which is a problem if a library is designed to solve a protocol once and for all :-). What happens right now if someone does send an exabyte sized frame at noio_ws?

Currently the deflate extension is the only one that wsproto supports ([I guess it's the only one that anyone supports](https://www.iana.org/assignments/websocket/websocket.xhtml#extension-name)), and it works incrementally just fine... for that matter I guess there are also SAX-style incremental JSON parsers, though I dunno how often people actually want to use them together with websockets.

**Comment by theelous3** [comment]
> What happens right now if someone does send an exabyte sized frame at noio_ws?

Everything dies unless you've an exabyte of ram :P 

Or you refuse it on the grounds of `1009 Message Too Big`, which is why that status code exists no?

```py
ws_conn = ws.Connection('CLIENT', max_buffer=126) # Bam! Protected.
```

It seems like handling it in another way (even if it's nice) is just getting around shortcomings in the protocol / bad services.

I really like the behavior of ``h11.Data`` which allows for dealing with bodies of unknown length, but in websockets everything is of known length. We can make informed decisions preemptively and control the limits applications work within.

**Comment by njsmith** [comment]
> in websockets everything is of known length

From a quick skim it looked to me like if someone sends a message fragmented into multiple frames, you wait for all the frames to arrive before returning anything? That case is pretty analogous to chunked encoding at the protocol level.

Anyway, I'm not trying to start an argument or anything; just thinking out loud about different API approaches. (h11 *also* has a max buffer argument, which is used to prevent obnoxiousness like exabyte-sized headers.)

**Comment by theelous3** [comment]
>Anyway, I'm not trying to start an argument or anything

On no of course. I'm absolutely delighted to hear criticisms and generally shoot the shit regarding websockets. As I said, I'm completely in a bubble here. It's nice to get some fresh feedback-air.

> That case is pretty analogous to chunked encoding at the protocol level.

You know, I hadn't thought about it like that. Earlier today I made a mental note of "why am I not returning `continue` frames as events, something is wrong". You've hit the nail on the head there.

Time to refactor!
